File: python/rapidsmpf/rapidsmpf/tests/streaming/test_channel_metadata.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES.
# SPDX-License-Identifier: Apache-2.0
"""Tests for streaming metadata types (Partitioning and ChannelMetadata)."""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

import pytest

import cudf
import pylibcudf as plc

from rapidsmpf.streaming.core.message import Message
from rapidsmpf.streaming.cudf import (
    ChannelMetadata,
    HashScheme,
    OrderKey,
    OrderScheme,
    Partitioning,
    TableChunk,
)
from rapidsmpf.utils.cudf import cudf_to_pylibcudf_table

if TYPE_CHECKING:
    from collections.abc import Sequence

    from rapidsmpf.streaming.core.context import Context


def test_hash_scheme() -> None:
    """Test HashScheme construction, properties, equality, and repr."""
    h1 = HashScheme((0, 1), 16)
    assert h1.column_indices == (0, 1)
    assert h1.modulus == 16
    assert repr(h1) == "HashScheme((0, 1), 16)"

    # Equality
    assert h1 == HashScheme((0, 1), 16)
    assert h1 != HashScheme((0, 1), 32)
    assert h1 != HashScheme((2,), 16)


def _two_key_order_scheme(*, strict_boundary: bool = False) -> OrderScheme:
    return OrderScheme(
        [
            OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE),
            OrderKey(1, plc.types.Order.DESCENDING, plc.types.NullOrder.AFTER),
        ],
        strict_boundary=strict_boundary,
    )


def test_order_key() -> None:
    k = OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE)
    assert k.column_index == 0
    assert k.order == plc.types.Order.ASCENDING
    assert k.null_order == plc.types.NullOrder.BEFORE
    assert k == OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE)
    assert k != OrderKey(1, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE)
    assert "OrderKey" in repr(k)

    with pytest.raises(ValueError, match="Invalid order"):
        OrderKey(0, 99, plc.types.NullOrder.BEFORE)  # type: ignore[arg-type]
    with pytest.raises(ValueError, match="Invalid null order"):
        OrderKey(0, plc.types.Order.ASCENDING, 99)  # type: ignore[arg-type]


def test_order_scheme() -> None:
    """Test OrderScheme construction, properties, equality, and repr."""
    o1 = _two_key_order_scheme()
    assert o1.keys == (
        OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE),
        OrderKey(1, plc.types.Order.DESCENDING, plc.types.NullOrder.AFTER),
    )
    assert o1.strict_boundary is False
    assert o1.get_boundaries_table() is None
    assert "OrderScheme" in repr(o1)

    # Equality
    assert o1 == _two_key_order_scheme()
    assert o1 != OrderScheme(
        [OrderKey(0, plc.types.Order.DESCENDING, plc.types.NullOrder.BEFORE)]
    )
    assert o1 != OrderScheme(
        [OrderKey(2, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE)]
    )

    o_strict = _two_key_order_scheme(strict_boundary=True)
    assert o_strict.strict_boundary is True
    assert o1 != o_strict
    assert o_strict == _two_key_order_scheme(strict_boundary=True)

    with pytest.raises(TypeError, match="OrderKey"):
        OrderScheme(
            cast(
                "Sequence[OrderKey]",
                [(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE)],
            )
        )

    with pytest.raises(ValueError, match="empty"):
        OrderScheme([])


def test_order_scheme_with_boundaries(context: Context) -> None:
    """Test OrderScheme with boundaries TableChunk (multi-column)."""
    df = cudf.DataFrame({"key1": [100, 200], "key2": ["abc", "xyz"]})
    stream = context.get_stream_from_pool()
    boundaries = TableChunk.from_pylibcudf_table(
        cudf_to_pylibcudf_table(df),
        stream,
        exclusive_view=False,
        br=context.br(),
    )
    o1 = OrderScheme(
        [
            OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE),
            OrderKey(1, plc.types.Order.DESCENDING, plc.types.NullOrder.AFTER),
        ],
        boundaries=boundaries,
    )
    assert o1.num_boundaries == 2
    tbl = o1.get_boundaries_table()
    assert tbl is not None
    assert tbl.num_columns() == 2
    assert tbl.num_rows() == 2


def test_num_boundaries() -> None:
    """num_boundaries is None without boundaries, returns row count otherwise."""
    o_no_b = _two_key_order_scheme()
    assert o_no_b.num_boundaries is None


def test_partitioning_scenarios() -> None:
    """Test various partitioning configurations."""
    # Default / None
    p_default = Partitioning()
    assert p_default.inter_rank is None
    assert p_default.local is None
    assert Partitioning(None, None) == p_default

    # Direct global shuffle: inter_rank=Hash, local=Aligned
    p_global = Partitioning(HashScheme((0,), 16), "inherit")
    assert p_global.inter_rank == HashScheme((0,), 16)
    assert p_global.local == "inherit"

    # Two-stage shuffle: inter_rank=Hash(nranks), local=Hash(N_l)
    p_twostage = Partitioning(HashScheme((0,), 4), HashScheme((0,), 8))
    assert p_twostage.inter_rank == HashScheme((0,), 4)
    assert p_twostage.local == HashScheme((0,), 8)

    # After local repartition: inter_rank=Hash, local=None
    p_local_none = Partitioning(HashScheme((0,), 16), None)
    assert p_local_none.inter_rank == HashScheme((0,), 16)
    assert p_local_none.local is None

    # Order-based partitioning (range partitioned / sorted)
    order_scheme = OrderScheme(
        [OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.AFTER)]
    )
    p_ordered = Partitioning(order_scheme, "inherit")
    assert p_ordered.inter_rank == order_scheme
    assert p_ordered.local == "inherit"

    # Mixed: inter_rank=Order, local=Hash
    p_mixed = Partitioning(
        OrderScheme(
            [OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE)]
        ),
        HashScheme((1,), 8),
    )
    assert isinstance(p_mixed.inter_rank, OrderScheme)
    assert isinstance(p_mixed.local, HashScheme)

    # Equality and repr
    assert p_global == Partitioning(HashScheme((0,), 16), "inherit")
    assert p_global != p_twostage
    assert "Partitioning" in repr(p_global)
    assert "inter_rank" in repr(p_global)

    # Invalid type
    with pytest.raises(TypeError):
        Partitioning("invalid", None)  # type: ignore[arg-type]


def test_channel_metadata() -> None:
    """Test ChannelMetadata construction and properties."""
    # Basic construction
    m = ChannelMetadata(local_count=4)
    assert m.local_count == 4
    assert m.duplicated is False

    p = Partitioning(HashScheme((0,), 16), "inherit")
    m_full = ChannelMetadata(local_count=4, partitioning=p, duplicated=True)
    assert m_full.partitioning == p
    assert m_full.duplicated is True

    # Equality and repr
    m2 = ChannelMetadata(local_count=4)
    assert m == m2
    assert m != ChannelMetadata(local_count=8)
    assert "local_count=4" in repr(m)

    # Validation
    with pytest.raises(ValueError, match="local_count must be non-negative"):
        ChannelMetadata(local_count=-1)


def test_message_roundtrip() -> None:
    """Test ChannelMetadata can round-trip through Message."""
    m = ChannelMetadata(
        local_count=4,
        partitioning=Partitioning(HashScheme((0,), 16), "inherit"),
        duplicated=True,
    )
    msg_m = Message(99, m)
    assert msg_m.sequence_number == 99
    got_m = ChannelMetadata.from_message(msg_m)
    assert got_m.local_count == 4
    assert got_m.duplicated is True
    assert got_m.partitioning.inter_rank == HashScheme((0,), 16)
    assert msg_m.empty()


def test_message_roundtrip_with_order_scheme(context: Context) -> None:
    """Test ChannelMetadata with OrderScheme can round-trip through Message."""
    df = cudf.DataFrame({"key1": [100, 200], "key2": ["abc", "xyz"]})
    stream = context.get_stream_from_pool()
    boundaries = TableChunk.from_pylibcudf_table(
        cudf_to_pylibcudf_table(df),
        stream,
        exclusive_view=False,
        br=context.br(),
    )
    order_scheme = OrderScheme(
        [
            OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE),
            OrderKey(1, plc.types.Order.DESCENDING, plc.types.NullOrder.AFTER),
        ],
        boundaries=boundaries,
        strict_boundary=True,
    )
    m = ChannelMetadata(
        local_count=8,
        partitioning=Partitioning(order_scheme, "inherit"),
        duplicated=True,
    )
    msg_m = Message(42, m)
    assert msg_m.sequence_number == 42
    got_m = ChannelMetadata.from_message(msg_m)
    assert got_m.local_count == 8
    assert got_m.duplicated is True
    assert isinstance(got_m.partitioning.inter_rank, OrderScheme)
    assert got_m.partitioning.inter_rank.keys == (
        OrderKey(0, plc.types.Order.ASCENDING, plc.types.NullOrder.BEFORE),
        OrderKey(1, plc.types.Order.DESCENDING, plc.types.NullOrder.AFTER),
    )
    assert got_m.partitioning.local == "inherit"
    assert got_m.partitioning.inter_rank.strict_boundary is True
    assert got_m.partitioning.inter_rank.num_boundaries == 2
    tbl = got_m.partitioning.inter_rank.get_boundaries_table()
    assert tbl is not None
    assert tbl.num_columns() == 2
    assert tbl.num_rows() == 2
    assert msg_m.empty()


def test_order_scheme_view_roundtrip() -> None:
    """Re-feeding a view OrderScheme (from metadata.partitioning.inter_rank) into
    Partitioning must use the live cpp_OrderScheme, not the wrapper's empty slot."""
    src = ChannelMetadata(
        local_count=1,
        partitioning=Partitioning(_two_key_order_scheme(), "inherit"),
    )
    # inter_rank is a non-owning view into src's storage
    view_scheme = src.partitioning.inter_rank
    assert isinstance(view_scheme, OrderScheme)

    # Round-trip the view through a new Partitioning
    p2 = Partitioning(view_scheme, None)
    assert p2.inter_rank == _two_key_order_scheme()


def test_access_after_move_raises() -> None:
    """Test that accessing a released ChannelMetadata raises ValueError."""
    m = ChannelMetadata(
        local_count=4,
        partitioning=Partitioning(HashScheme((0,), 16), "inherit"),
    )
    # Move into a message (releases the handle)
    _ = Message(0, m)

    # Accessing any property should raise ValueError
    with pytest.raises(ValueError, match="uninitialized"):
        _ = m.local_count

    with pytest.raises(ValueError, match="uninitialized"):
        _ = m.partitioning

    with pytest.raises(ValueError, match="uninitialized"):
        _ = m.duplicated

    with pytest.raises(ValueError, match="uninitialized"):
        repr(m)

File: cpp/tests/streaming/test_channel_metadata.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES. All rights
 * reserved. SPDX-License-Identifier: Apache-2.0
 */

#include <stdexcept>

#include <gtest/gtest.h>

#include <rapidsmpf/streaming/cudf/channel_metadata.hpp>

using namespace rapidsmpf::streaming;

namespace {

OrderScheme make_order_scheme_two_key(bool strict_boundaries) {
    return OrderScheme{
        {{0, cudf::order::ASCENDING, cudf::null_order::BEFORE},
         {1, cudf::order::DESCENDING, cudf::null_order::AFTER}},
        nullptr,
        strict_boundaries,
    };
}

}  // namespace

class StreamingChannelMetadata : public ::testing::Test {};

TEST_F(StreamingChannelMetadata, HashScheme) {
    HashScheme h{{0, 1}, 16};
    EXPECT_EQ(h.column_indices.size(), 2);
    EXPECT_EQ(h.column_indices[0], 0);
    EXPECT_EQ(h.column_indices[1], 1);
    EXPECT_EQ(h.modulus, 16);

    // Equality
    EXPECT_EQ(h, (HashScheme{{0, 1}, 16}));
    EXPECT_NE(h, (HashScheme{{0, 1}, 32}));
    EXPECT_NE(h, (HashScheme{{2}, 16}));
}

TEST_F(StreamingChannelMetadata, OrderScheme) {
    OrderScheme o{
        {{0, cudf::order::ASCENDING, cudf::null_order::BEFORE},
         {1, cudf::order::DESCENDING, cudf::null_order::AFTER}},
        nullptr
    };
    EXPECT_EQ(o.keys.size(), 2);
    EXPECT_EQ(o.keys[0].column_index, 0);
    EXPECT_EQ(o.keys[0].order, cudf::order::ASCENDING);
    EXPECT_EQ(o.keys[0].null_order, cudf::null_order::BEFORE);
    EXPECT_EQ(o.keys[1].column_index, 1);
    EXPECT_EQ(o.keys[1].order, cudf::order::DESCENDING);
    EXPECT_EQ(o.keys[1].null_order, cudf::null_order::AFTER);
    EXPECT_EQ(o.boundaries, nullptr);
    EXPECT_FALSE(o.strict_boundaries);

    OrderScheme o_strict{
        {{0, cudf::order::ASCENDING, cudf::null_order::BEFORE},
         {1, cudf::order::DESCENDING, cudf::null_order::AFTER}},
        nullptr,
        true,
    };
    EXPECT_NE(o, o_strict);
    EXPECT_TRUE(o_strict.strict_boundaries);

    OrderScheme o_same{
        {{0, cudf::order::ASCENDING, cudf::null_order::BEFORE},
         {1, cudf::order::DESCENDING, cudf::null_order::AFTER}},
        nullptr
    };
    EXPECT_EQ(o, o_same);

    EXPECT_NE(
        o, (OrderScheme{{{2, cudf::order::ASCENDING, cudf::null_order::BEFORE}}, nullptr})
    );
    EXPECT_NE(
        o,
        (OrderScheme{
            {{0, cudf::order::DESCENDING, cudf::null_order::BEFORE},
             {1, cudf::order::DESCENDING, cudf::null_order::AFTER}},
            nullptr
        })
    );
    EXPECT_NE(
        o,
        (OrderScheme{
            {{0, cudf::order::ASCENDING, cudf::null_order::AFTER},
             {1, cudf::order::DESCENDING, cudf::null_order::AFTER}},
            nullptr
        })
    );
}

TEST_F(StreamingChannelMetadata, FromOrderRejectsEmptyKeys) {
    EXPECT_THROW(
        static_cast<void>(PartitioningSpec::from_order(OrderScheme{{}, nullptr})),
        std::invalid_argument
    );
}

TEST_F(StreamingChannelMetadata, PartitioningSpec) {
    // None
    auto spec_none = PartitioningSpec::none();
    EXPECT_EQ(spec_none.type, PartitioningSpec::Type::NONE);

    // Inherit
    auto spec_inherit = PartitioningSpec::inherit();
    EXPECT_EQ(spec_inherit.type, PartitioningSpec::Type::INHERIT);

    // Hash
    auto spec_hash = PartitioningSpec::from_hash(HashScheme{{0}, 16});
    EXPECT_EQ(spec_hash.type, PartitioningSpec::Type::HASH);
    EXPECT_EQ(spec_hash.hash->column_indices[0], 0);
    EXPECT_EQ(spec_hash.hash->modulus, 16);

    // Order
    auto spec_order = PartitioningSpec::from_order(
        OrderScheme{{{0, cudf::order::ASCENDING, cudf::null_order::BEFORE}}, nullptr}
    );
    EXPECT_EQ(spec_order.type, PartitioningSpec::Type::ORDER);
    EXPECT_EQ(spec_order.order->keys[0].column_index, 0);
    EXPECT_EQ(spec_order.order->keys[0].order, cudf::order::ASCENDING);
    EXPECT_EQ(spec_order.order->keys[0].null_order, cudf::null_order::BEFORE);

    // Equality
    EXPECT_EQ(spec_none, PartitioningSpec::none());
    EXPECT_EQ(spec_inherit, PartitioningSpec::inherit());
    EXPECT_EQ(spec_hash, PartitioningSpec::from_hash(HashScheme{{0}, 16}));
    EXPECT_NE(spec_none, spec_inherit);
    EXPECT_NE(spec_hash, PartitioningSpec::from_hash(HashScheme{{0}, 32}));
    EXPECT_NE(spec_hash, spec_order);

    auto spec_order_same = PartitioningSpec::from_order(
        OrderScheme{{{0, cudf::order::ASCENDING, cudf::null_order::BEFORE}}, nullptr}
    );
    EXPECT_EQ(spec_order, spec_order_same);
}

TEST_F(StreamingChannelMetadata, PartitioningScenarios) {
    // Default construction
    Partitioning p_default{};
    EXPECT_EQ(p_default.inter_rank.type, PartitioningSpec::Type::NONE);
    EXPECT_EQ(p_default.local.type, PartitioningSpec::Type::NONE);

    // Direct global shuffle: inter_rank=Hash, local=Inherit
    Partitioning p_global{
        PartitioningSpec::from_hash(HashScheme{{0}, 16}), PartitioningSpec::inherit()
    };
    EXPECT_EQ(p_global.inter_rank.type, PartitioningSpec::Type::HASH);
    EXPECT_EQ(p_global.local.type, PartitioningSpec::Type::INHERIT);
    EXPECT_EQ(p_global.inter_rank.hash->modulus, 16);

    // Two-stage shuffle: inter_rank=Hash(nranks), local=Hash(N_l)
    Partitioning p_twostage{
        PartitioningSpec::from_hash(HashScheme{{0}, 4}),
        PartitioningSpec::from_hash(HashScheme{{0}, 8})
    };
    EXPECT_EQ(p_twostage.inter_rank.hash->modulus, 4);
    EXPECT_EQ(p_twostage.local.hash->modulus, 8);

    // Order-based partitioning (range partitioned / sorted)
    Partitioning p_ordered{
        PartitioningSpec::from_order(
            OrderScheme{{{0, cudf::order::ASCENDING, cudf::null_order::AFTER}}, nullptr}
        ),
        PartitioningSpec::inherit()
    };
    EXPECT_EQ(p_ordered.inter_rank.type, PartitioningSpec::Type::ORDER);
    EXPECT_EQ(p_ordered.local.type, PartitioningSpec::Type::INHERIT);
    EXPECT_EQ(p_ordered.inter_rank.order->keys[0].column_index, 0);

    // Mixed: inter_rank=Order, local=Hash
    Partitioning p_mixed{
        PartitioningSpec::from_order(
            OrderScheme{{{0, cudf::order::DESCENDING, cudf::null_order::BEFORE}}, nullptr}
        ),
        PartitioningSpec::from_hash(HashScheme{{1}, 8})
    };
    EXPECT_EQ(p_mixed.inter_rank.type, PartitioningSpec::Type::ORDER);
    EXPECT_EQ(p_mixed.local.type, PartitioningSpec::Type::HASH);

    // Equality
    EXPECT_EQ(
        p_global,
        (Partitioning{
            PartitioningSpec::from_hash(HashScheme{{0}, 16}), PartitioningSpec::inherit()
        })
    );
    EXPECT_NE(p_global, p_twostage);
    EXPECT_NE(p_global, p_ordered);
}

TEST_F(StreamingChannelMetadata, ChannelMetadata) {
    // Full construction - use std::move to avoid GCC false positive on vector copy
    Partitioning p{
        PartitioningSpec::from_hash(HashScheme{{0}, 16}), PartitioningSpec::inherit()
    };
    ChannelMetadata m{4, std::move(p), true};
    EXPECT_EQ(m.local_count, 4);
    EXPECT_EQ(m.partitioning.inter_rank.type, PartitioningSpec::Type::HASH);
    EXPECT_EQ(m.partitioning.local.type, PartitioningSpec::Type::INHERIT);
    EXPECT_TRUE(m.duplicated);

    // Minimal construction
    ChannelMetadata m_minimal{4};
    EXPECT_EQ(m_minimal.local_count, 4);
    EXPECT_FALSE(m_minimal.duplicated);

    // Equality - create fresh partitionings and move them
    ChannelMetadata m_same{
        4,
        Partitioning{
            PartitioningSpec::from_hash(HashScheme{{0}, 16}), PartitioningSpec::inherit()
        },
        true
    };
    ChannelMetadata m_diff{
        8,
        Partitioning{
            PartitioningSpec::from_hash(HashScheme{{0}, 16}), PartitioningSpec::inherit()
        },
        true
    };
    EXPECT_EQ(m, m_same);
    EXPECT_NE(m, m_diff);
}

TEST_F(StreamingChannelMetadata, MessageRoundTrip) {
    // ChannelMetadata round-trip with HashScheme
    Partitioning part{
        PartitioningSpec::from_hash(HashScheme{{0}, 16}), PartitioningSpec::inherit()
    };
    auto m = std::make_unique<ChannelMetadata>(4, std::move(part), false);
    auto msg_m = to_message(99, std::move(m));
    EXPECT_EQ(msg_m.sequence_number(), 99);
    EXPECT_TRUE(msg_m.holds<ChannelMetadata>());
    auto released = msg_m.release<ChannelMetadata>();
    EXPECT_EQ(released.local_count, 4);
    EXPECT_FALSE(released.duplicated);
    EXPECT_EQ(released.partitioning.inter_rank.hash->modulus, 16);
    EXPECT_TRUE(msg_m.empty());
}

TEST_F(StreamingChannelMetadata, MessageRoundTripWithOrderScheme) {
    // ChannelMetadata round-trip with OrderScheme (full field assertions).
    Partitioning part{
        PartitioningSpec::from_order(make_order_scheme_two_key(false)),
        PartitioningSpec::inherit()
    };
    auto m = std::make_unique<ChannelMetadata>(8, std::move(part), true);
    auto msg_m = to_message(42, std::move(m));
    EXPECT_EQ(msg_m.sequence_number(), 42);
    EXPECT_TRUE(msg_m.holds<ChannelMetadata>());
    auto released = msg_m.release<ChannelMetadata>();
    EXPECT_EQ(released.local_count, 8);
    EXPECT_TRUE(released.duplicated);
    EXPECT_EQ(released.partitioning.inter_rank.type, PartitioningSpec::Type::ORDER);
    EXPECT_EQ(released.partitioning.inter_rank.order->keys.size(), 2);
    EXPECT_EQ(released.partitioning.inter_rank.order->keys[0].column_index, 0);
    EXPECT_EQ(
        released.partitioning.inter_rank.order->keys[0].order, cudf::order::ASCENDING
    );
    EXPECT_EQ(
        released.partitioning.inter_rank.order->keys[0].null_order,
        cudf::null_order::BEFORE
    );
    EXPECT_EQ(released.partitioning.inter_rank.order->keys[1].column_index, 1);
    EXPECT_EQ(
        released.partitioning.inter_rank.order->keys[1].order, cudf::order::DESCENDING
    );
    EXPECT_EQ(
        released.partitioning.inter_rank.order->keys[1].null_order,
        cudf::null_order::AFTER
    );
    EXPECT_EQ(released.partitioning.local.type, PartitioningSpec::Type::INHERIT);
    EXPECT_FALSE(released.partitioning.inter_rank.order->strict_boundaries);
    EXPECT_TRUE(msg_m.empty());
}

TEST_F(StreamingChannelMetadata, MessageRoundTripWithOrderSchemeStrictBoundary) {
    // Only strict_boundaries + Message path; full OrderScheme field checks are above.
    Partitioning part{
        PartitioningSpec::from_order(make_order_scheme_two_key(true)),
        PartitioningSpec::inherit()
    };
    auto m = std::make_unique<ChannelMetadata>(8, std::move(part), false);
    auto msg_m = to_message(43, std::move(m));
    auto released = msg_m.release<ChannelMetadata>();
    EXPECT_TRUE(released.partitioning.inter_rank.order->strict_boundaries);
}

TEST_F(StreamingChannelMetadata, OrderSchemeCopySharesBoundaries) {
    // Copying an OrderScheme shares the shared_ptr boundary object (not a deep copy).
    OrderScheme original{make_order_scheme_two_key(false)};
    OrderScheme copy = original;
    EXPECT_EQ(copy, original);
    EXPECT_EQ(
        copy.boundaries, original.boundaries
    );  // same shared_ptr (both nullptr here)
    // Keys vector is independently owned; mutating the copy does not affect original.
    copy.keys.clear();
    EXPECT_FALSE(original.keys.empty());
}
